clang/test/SemaSYCL/sycl-device-const-static.cpp

```diff
@@ -51,13 +51,34 @@ void usage() {
 
 template <typename Name, typename Func>
 __attribute__((sycl_kernel)) void kernel_single_task(const Func &kernelFunc) {
+  // expected-error@+1{{SYCL kernel cannot use a non-const static data variable}}
```

Comment:
Oh, that is interesting. The message says "cannot use", but it is not used, just declared here. Do you know what it the reason for error emission here?

Reply:
This is a good question.   The spec says that `"Variables with static storage duration that are odr-used inside a device function, must be const or constexpr and zero-initialized or constant-initialized."`
Since this is a definition of a static variable (which also initializes it), I am not entirely sure if it also meets the rule of "odr-use".   So it is possible that we can let this case slide until it is actually used in the device code; at that point it is definitely an error, so I am not too unhappy that we issue a diagnostic at definition time.   If you like, I can add a FIXME to this part of the test.

Note that the diagnostic itself was added PR #1141 and later extended to also cover thread-locals in PR #1281.   So the case has been diagnosed for a while now.   Looks like PR #5195  you removed it, may be because you felt that it was not used?  So adding this diagnostic back is only a recent change.

Reply:
I see. It seems #1141 added it this way, so the declaration (not the use) is prohibited. Probably worth checking if the modern spec disallows it and if it is, update error message for better user experience (now it is a bit confusing). That is out of the scope of your patch, I think.